Toolbox/arcsdm/csi/core/calculation.py

```diff
@@ -0,0 +1,119 @@
+import arcpy
+import numpy as np
+from typing import List, Optional
+from arcsdm.csi.core.calculate_evidence_matrix import calculate_evidence_matrix
+from arcsdm.csi.core.corner_csi import calculate_corner_csi
+from arcsdm.csi.core.centroid_csi import calculate_centroid_to_centroid_csi
+from arcsdm.csi.analysis.class_centroids import calculate_class_centroids
+from arcsdm.csi.analysis.pixel_to_label_csi import pixel_to_label_csi
+
+from arcsdm.csi.helpers.save_results import save_csv_results
+from arcsdm.csi.helpers.rows_with_labels import rows_with_labels
+from arcsdm.csi.helpers.load_data import (load_labeled_data, load_raster_data)
+
+def calculation(
+    selected_label_field: Optional[str],
+    labelled_path: str,
+    label_field_names: List[str],
+    coordinate_field_names: List[str],
+    feature_field_names: List[str],
+    rasters_list: List[str],
+    evidence_type: Optional[str],
+    csv_nodata: float,
+    out_labelled_pairwise_csv: str,
+    out_evidence_matrix_csv: Optional[str],
+    out_individual_evidence_csv: Optional[str],
+    out_raster_folder: Optional[str],
+    out_class_centroid: Optional[str] = None,
+    out_centroid_csi_csv: Optional[str] = None,
+):
+    """Perform CSI calculation workflow."""
+    # Load labeled data
+    all_df, feature_fields, has_geometry = load_labeled_data(
+        labelled_path, label_field_names, feature_field_names, coordinate_field_names
+    )
+
+    if all_df is None:
+        arcpy.AddError("Failed to load labeled data - cannot proceed with analysis.")
+        return
+
+    # Filter to only labeled points
+    label_mask = rows_with_labels(all_df, label_field_names, csv_nodata)
+    labeled_df = all_df.loc[label_mask].reset_index(drop=True)
+
+    # Further filter by selected label field
+    if selected_label_field and selected_label_field in labeled_df.columns:
+        selected_mask = labeled_df[selected_label_field].notna()
+        labeled_df = labeled_df.loc[selected_mask].reset_index(drop=True)
+        arcpy.AddMessage(f"Filtered to {len(labeled_df)} points with valid {selected_label_field}")
+
+    arcpy.AddMessage(f"Using {len(labeled_df)} labeled points for analysis")
+
+    if len(labeled_df) == 0:
+        arcpy.AddError("No labeled rows found - cannot proceed with analysis.")
+        return
+
+    # Calculate corner CSI matrix
+    arcpy.AddMessage("\nCorner CSI Matrix Calculation")
+    corner_matrix = calculate_corner_csi(labeled_df, feature_fields, csv_nodata)
+
+    # Calculate class centroids
+    centroids_df = calculate_class_centroids(
```

Comment:
If the user has not added an output centroids.csv, there's no need to calculate the centered values.